### PR TITLE
[sdk] Use reflection and dynamic types to start providers

### PR DESCRIPTION
**Pull request by CodeBlanch**
I've received a lot of feedback that `StartWithHost` isn't great 🤣 Chatted offline with some folks who seemed to think a reflection-based solution would be better so here goes...

## Changes

* Removes `StartWithHost` and uses magic to let the `AddOpenTelemetry()` extension (part of SDK) own that responsibility without a dependency on `Microsoft.Extensions.Hosting.Abstractions`.

## Details

With this change the `IHostedService` currently registered into `IServiceCollection` by `OpenTelemetry.Extensions.Hosting` becomes part of the SDK. If the SDK finds the `Microsoft.Extensions.Hosting.IHostedService` type (reflectively) and `RuntimeFeature.IsDynamicCodeSupported` is true the SDK will automatically build an `IHostedService` and drop it into the `IServiceCollection`. We won't need the `OpenTelemetry.Extensions.Hosting` package at all if we do this... we can just mark it deprecated and remove the code from the solution. Users who do not have dynamic code support can either access the `TracerProvider` and/or `MeterProvider` manually from the `IServiceProvider` or use the "Sdk.Create*().Build()" configuration style (aka "Console" style).

## TODOs

* [X] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [X] Fix up tests
~* [ ] Fix up documentation~ Will do this as a follow-up.
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by samsp-msft** [comment]
Relying on ref emit is going to cause problems with Native AOT and potentially trimming. While there will need to be application code differences for those apps - we want to keep that to a minimum, and I would suggest that this is such a fundamental scenario that having differences is going to be bad.

**Comment by CodeBlanch** [comment]
@samsp-msft What's funny is I spent much effort in .NET 7 and OTel 1.4 to remove our reflection emit only to put it back for this 🤣 We could add a dependency on `Microsoft.Extensions.Hosting.Abstractions` in the SDK to not need it, but I don't think many people would go for that. Plus it only works when an actual host is being used. .NET Framework, Console apps, it is just noise. I think we need to view this as an interim step. My hope was to work with the .NET team on an API for .NET 8 we could use to warm up the SDK and then we could remove this. Thoughts on that path?

**Review by reyang — APPROVED** [review]
LGTM. This is a nice user experience improvement / simplification. We need to follow up with .NET runtime folks and see how to eventually get rid of reflection (not blocking).

File: src/OpenTelemetry.Extensions.Hosting/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+* Removed the `OpenTelemetryBuilder.StartWithHost` extension and moved the
```
Comment (cijothomas) [inline src/OpenTelemetry.Extensions.Hosting/CHANGELOG.md:5]: Lets make it here itself that, with this change, this whole package itself will not be required anymore, and will be deprecated with no further release.

Comment (reyang) [reply src/OpenTelemetry.Extensions.Hosting/CHANGELOG.md:5]: +1

File: src/OpenTelemetry/CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,10 @@
 * Removed the dependency on System.Reflection.Emit.Lightweight
   ([#4140](https://github.com/open-telemetry/opentelemetry-dotnet/pull/4140))
 
+* The `AddOpenTelemetry` extension will now register an `IHostedService` if
```
Comment (cijothomas) [inline src/OpenTelemetry/CHANGELOG.md:8]: lets do a more detailed write up here for folks updating to this version. Most end users do not really have the context about this, so we should write the steps one should follow when updating to this version.
1. Remove the nuget reference to otel.ext.hosting.
2. Remove the startwithhost(). 
3. (change ns based on the other PR from alan.)

Its okay to do this in a separate PR.

Comment (cijothomas) [reply src/OpenTelemetry/CHANGELOG.md:8]: also need to write the steps for those users who have to manually retrieve MP/TP .

File: src/OpenTelemetry/Internal/HostingHelper.cs [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using OpenTelemetry.Metrics;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Internal;
+
+internal static class HostingHelper
+{
+    private static readonly object LockObject = new();
```
Comment (cijothomas) [inline src/OpenTelemetry/Internal/HostingHelper.cs:32]:
```suggestion
    private static readonly object lockObject = new();
```

Comment (cijothomas) [reply src/OpenTelemetry/Internal/HostingHelper.cs:32]: nvm

File: src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs [hunk 4]
```diff
@@ -415,6 +424,36 @@ public void InvalidEnvironmentVariable(string key, string value)
             this.WriteEvent(47, key, value);
         }
 
+        [Event(48, Message = "OpenTelemetry IHostedService registered in application services.", Level = EventLevel.Informational)]
+        public void HostedServiceRegistered()
+        {
+            this.WriteEvent(48);
+        }
+
+        [Event(49, Message = "OpenTelemetry IHostedService application services registration skipped. Reason: '{0}'", Level = EventLevel.Informational)]
```
Comment (cijothomas) [inline src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs:433]: Warning? This is user-actionable right? We could point users to the steps they should take to resolve this.

File: src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs [hunk 5]
```diff
@@ -415,6 +424,36 @@ public void InvalidEnvironmentVariable(string key, string value)
...
+        public void HostedServiceRegistrationSkipped(string reason)
+        {
+            this.WriteEvent(49, reason);
+        }
+
+        [Event(50, Message = "OpenTelemetry IHostedService could not be registered in application services. Error: '{0}'", Level = EventLevel.Warning)]
```
Comment (cijothomas) [inline src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs:439]: this can be Error?

**Comment by dpk83** [comment]
I don't think using reflection here is a clean approach. Why can't we have the existing AddOpenTelemetryMetrics and AddOpenTelemetryTracing APIs in Extensions.Hosting. 

The Extensions.Hosting package is a different package that provides a better DI support model so this package exposing different APIs is better than using the hacky reflection here in my opinion.

File: docs/trace/getting-started-aspnetcore/Program.cs [hunk 6]
```diff
@@ -27,8 +27,7 @@
         .AddService(serviceName: "OTel.NET Getting Started"))
     .WithTracing(builder => builder
         .AddAspNetCoreInstrumentation()
-        .AddConsoleExporter())
-    .StartWithHost();
+        .AddConsoleExporter());
```
Comment (cijothomas) [inline docs/trace/getting-started-aspnetcore/Program.cs:30]: I am not able to recollect the behavior.. could you help clarify the behavior when something fails ? Does this throw and allow fail-fast? Or we swallow exceptions and log, and proceed?

File: src/OpenTelemetry/Internal/HostingHelper.cs [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+        if (!isDynamicCodeSupported)
+        {
+            reason = "Dynamic code not supported";
+            return false;
+        }
+
+        var iHostedServiceType = Type.GetType(
+            "Microsoft.Extensions.Hosting.IHostedService, Microsoft.Extensions.Hosting.Abstractions", throwOnError: false);
+
+        if (iHostedServiceType == null)
```
Comment (dpk83) [inline src/OpenTelemetry/Internal/HostingHelper.cs:65]: How will it behave for AzureFunctions. I believe IHostedService will be found on AzureFunctions but the hostedService infra is not supported.

Comment (cijothomas) [reply src/OpenTelemetry/Internal/HostingHelper.cs:65]: I think Functions team added special support for OTel Hosted Service. I'll try to find that PR, but yes this is something to be tried out. (Independent of this, Otel 1.4 wont work in Functions unless dedicated plan, due to Functions not supporting DS 7.0)

**Comment by cijothomas** [comment]
Summarizing my understanding/thoughts:

With 1.4.0-rc3, OTel SDK has native DI capabilities.
We still needed Otel.Ext.Hosting package, purely to add a IHostedService to DI. The HostedService exists only to trigger retrieving Tracer/Meter Provider, which in turn, trigger building the whole SDK pipeline. 
If there was something from DI that could be used instead to trigger building of pipeline, we wouldn't have the need for this IHostedService trick. Such capability do not exist today, and we hope to work with .NET team to get it added in future. Until that occurs, this reflection hack seems a reasonable option.

Also, we need to write a clear step-by-step guide for users upgrading to this version as an issue, and link to that issue from the package deprecations, Obsolete messages, and possibly from internal logs. I can imagine a lot of users (who are not actively following this), would be confused when the package goes missing, and all the breaking changes.

File: src/OpenTelemetry/Internal/HostingHelper.cs (hunk 7 above)
Comment (dpk83) [reply src/OpenTelemetry/Internal/HostingHelper.cs:65]: So, Otel is going to break for Azure functions with 1.4 upgrade? Shouldn't there be a major version bump if it's breaking the functionality for a group of customers?

**Comment by dpk83** [comment]
> Such capability do not exist today, and we hope to work with .NET team to get it added in future. Until that occurs, this reflection hack seems a reasonable option.

When such ability gets into .NET, is it going to be in all .NET versions? If not than this reflection is here to live for a very long time, right?

> Also, we need to write a clear step-by-step guide for users upgrading to this version as an issue, and link to that issue from the package deprecations, Obsolete messages, and possibly from internal logs. I can imagine a lot of users (who are not actively following this), would be confused when the package goes missing, and all the breaking changes.

It's a breaking change so wouldn't it demand a major version bump?

File: src/OpenTelemetry/Internal/HostingHelper.cs (hunk 7 above)
Comment (cijothomas) [reply src/OpenTelemetry/Internal/HostingHelper.cs:65]: I do not think so. There is no breaking change in 1.3 -> 1.4. Its just Azure Functions has a strict DS version dependency. As mentioned in the issue in Azure Functions host, the issue is not with OTel. Any app with DS 7.0 will not work.
https://github.com/Azure/azure-functions-host/issues/8938

(IIRC, this issue was there from 1.0 of OTel onwards (https://github.com/Azure/azure-functions-host/issues/7135) and it got fixed, but again broken (with every new .NET release). Hopefully by the time 1.4 is stable, Functions will fix it :)

Either way, lets use https://github.com/open-telemetry/opentelemetry-dotnet/issues/4047 for this conversation, as this is unrelated to this PR.

Comment (dpk83) [reply src/OpenTelemetry/Internal/HostingHelper.cs:65]: > Independent of this, Otel 1.4 wont work in Functions unless dedicated plan, due to Functions not supporting DS 7.0

For OpenTelemetry .NET customer's point of view it's a breaking change. i.e. Customer using 1.3 upgrades to 1.4, they won't expect any functionality to break or any breaking changes at compile time because this isn't a major version upgrade. So, from OpenTelemetry .NET library's upgrading to DS 7.0 is resulting in a breaking change for it's customer.

**Comment by cijothomas** [comment]
> > Also, we need to write a clear step-by-step guide for users upgrading to this version as an issue, and link to that issue from the package deprecations, Obsolete messages, and possibly from internal logs. I can imagine a lot of users (who are not actively following this), would be confused when the package goes missing, and all the breaking changes.
> 
> It's a breaking change so wouldn't it demand a major version bump?

There is no breaking change in any stable releases. Only in pre-release versions, so don't warrant major version bump.

File: src/OpenTelemetry/Internal/HostingHelper.cs (hunk 7 above)
Comment (cijothomas) [reply src/OpenTelemetry/Internal/HostingHelper.cs:65]: >  Customer using 1.3 upgrades to 1.4, they won't expect any functionality to break or any breaking changes at compile time

^ There is none.

Comment (dpk83) [reply src/OpenTelemetry/Internal/HostingHelper.cs:65]: > > Independent of this, Otel 1.4 wont work in Functions unless dedicated plan, due to Functions not supporting DS 7.0
> 
> For OpenTelemetry .NET customer's point of view it's a breaking change. i.e. Customer using 1.3 upgrades to 1.4, they won't expect any functionality to break or any breaking changes at compile time because this isn't a major version upgrade. So, from OpenTelemetry .NET library's upgrading to DS 7.0 is resulting in a breaking change for it's customer.

This I am saying not only for Azure functions but for all other customers

Comment (cijothomas) [reply src/OpenTelemetry/Internal/HostingHelper.cs:65]: There is no breaking change for all customers!

Comment (dpk83) [reply src/OpenTelemetry/Internal/HostingHelper.cs:65]: If the issue is only with azure functions than it's okay. If it hits any regular .NET flavors than it's a major breaking change.

**Comment by samsp-msft** [comment]
> @samsp-msft What's funny is I spent much effort in .NET 7 and OTel 1.4 to remove our reflection emit only to put it back for this 🤣 We could add a dependency on `Microsoft.Extensions.Hosting.Abstractions` in the SDK to not need it, but I don't think many people would go for that. Plus it only works when an actual host is being used. .NET Framework, Console apps, it is just noise. I think we need to view this as an interim step. My hope was to work with the .NET team on an API for .NET 8 we could use to warm up the SDK and then we could remove this. Thoughts on that path?

I would suggest that taking a dependency on Microsoft.Extensions.Hosting.Abstractions from OpenTelemetry.Extensions.DependencyInjection would probably be cleaner and have less impact on customers than this change. There is going to be a big push for Native AOT for ASP.NET Core, and so this will be the wrong direction at that time,

File: src/OpenTelemetry/Internal/HostingHelper.cs (hunk 7 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Internal/HostingHelper.cs:65]: This is a great callout. Shamefully, I was not aware of this in Azure Functions! Playing with the code right now, I think we can make it backwards compatible with what functions is doing.

Comment (CodeBlanch) [reply src/OpenTelemetry/Internal/HostingHelper.cs:65]: OK I just pushed an update that makes Azure Functions happy locally. Need to verify on an actual build of the packages but I think we're good.

Comment (cijothomas) [reply src/OpenTelemetry/Internal/HostingHelper.cs:65]: @RohitRanjanMS Would you be able to help confirm if this will continue to work in Functions?

File: src/OpenTelemetry/CHANGELOG.md (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/CHANGELOG.md:8]: I will tackle this in a follow-up. Need to update documentation and then I'll drop a link here for more info.

File: src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs (hunk 5 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs]: Switched to Error

**Review by cijothomas — APPROVED** [review]
some follow ups required in docs to help users migrate smoothly, but this should not block this PR.

File: src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs (hunk 4 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Internal/OpenTelemetrySdkEventSource.cs]: There are 3 reasons we could log skipped:

1) `RuntimeFeature.IsDynamicCodeSupported` == `false`
2) `Microsoft.Extensions.Hosting.IHostedService, Microsoft.Extensions.Hosting.Abstractions` type could not be found
3) Reflection/dynamic code blew up. This will also cause an Error to be written for `HostedServiceRegistrationFailure` event.

1 + 2 could be normal/expected, just depends on who is setting up the host. I just updated it to be a warning but if any of that changed your mind LMK.

File: docs/trace/getting-started-aspnetcore/Program.cs (hunk 6 above)
Comment (CodeBlanch) [reply docs/trace/getting-started-aspnetcore/Program.cs:30]: As of https://github.com/open-telemetry/opentelemetry-dotnet/pull/4006 we will throw from the `IHostedService` `StartAsync` method if something throws during "build" phase.

**Review by alanwest — APPROVED** [review]

File: src/OpenTelemetry/Internal/HostingHelper.cs (hunk 7 above)
Comment (RohitRanjanMS) [reply src/OpenTelemetry/Internal/HostingHelper.cs:65]: Thanks @cijothomas, yes, this will work for Functions.

**Comment by hejingkan2005** [comment]
@CodeBlanch , @cijothomas , I meet the same issue after OTel migration to 1.4.0-rc.3, may I confirm with you whether this PR is the fix for the issue - "Could not load file or assembly 'System.Diagnostics.DiagnosticSource'" with Azure function? What's the version of OTel is going to release?

**Comment by Kielek** [comment]
@hejingkan2005, please check 1.4.0-rc.4.

**Comment by hejingkan2005** [comment]
@Kielek , thank you for your comment, but 1.4.0-rc.4 is not applicable as we get two other dependent packages which ask for version must equal to 1.4.0-rc.3:
+++++++++++++++
\SearchCrawler.WorkerFunction.csproj : error NU1608: Detected package version outside of dependency constraint: OpenTelemetry.Exporter.Geneva 1.4.0-rc.3 requires OpenTelemetry (= 1.4.0-rc.3) but version OpenTelemetry 1.4.0-rc.4 was resolved.
\SearchCrawler.WorkerFunction\SearchCrawler.WorkerFunction.csproj : error NU1608: Detected package version outside of dependency constraint: Azure.Monitor.OpenTelemetry.Exporter 1.0.0-beta.7 requires OpenTelemetry (= 1.4.0-rc.3) but version OpenTelemetry 1.4.0-rc.4 was resolved.
+++++++++++++++

**Comment by cijothomas** [comment]
> @CodeBlanch , @cijothomas , I meet the same issue after migrating OTel to 1.4.0-rc.3, may I confirm with you whether this PR is the fix for the issue - "Could not load file or assembly 'System.Diagnostics.DiagnosticSource, version=7.0.0 ...'" with Azure function? What's the version of OTel is going to release?

This PR (which was reverted), is not for addressing any Functions specific issue.
